Add new application definition for PEEM experiments

PeterC-DLS opened the pull request:
This extends NXmpes and adds some new groups and fields and also adjusts some existing fields' classes.

NB this is a rebase of #65

PeterC-DLS committed:
Add new application definition for PEEM experiments

This extends NXmpes and adds some new groups and fields and also adjusts some existing fields' classes

PeterC-DLS committed:
Update mpes_peem definition

Move YAML to nyaml directory, tweak PEEM definition according to feedback, and add some units

PeterC-DLS committed:
Update mpes_peem defintion

Fix some indentations, switch to unit, add an NXsource for PEEM case, and
rename coordinate_system_set to geometries

lukaspie commented on contributed_definitions/NXmpes_peem.nxdl.xml:
```diff
@@ -0,0 +1,202 @@
+                    </field>
+                </group>
+            </group>
+            <group type="NXoptical_system_em">
```
We could potentially add this as optional to `NXmpes` as well

lukaspie commented on contributed_definitions/NXmpes_peem.nxdl.xml:
```diff
@@ -0,0 +1,202 @@
+                            <item value="LEEM dark-field"/>
+                            <item value="LEED"/>
+                            <item value="dispersive plane"/>
+                            <item value="momentum microscope"/>
```
Should any of these new items be added to `NXmpes`?

This is the current list there:
<img width="245" height="112" alt="image" src="https://github.com/user-attachments/assets/4f3054da-aa56-4e4b-9c1a-38da4492f26b" />

PeterC-DLS replied:
Done

lukaspie commented on contributed_definitions/NXmpes_peem.nxdl.xml:
```diff
@@ -0,0 +1,202 @@
+                        <enumeration>
+                            <item value="real"/>
+                            <item value="reciprocal"/>
+                            <item value="energy"/>
```
To be added to the base class?

lukaspie commented on contributed_definitions/NXmpes_peem.nxdl.xml:
```diff
@@ -0,0 +1,202 @@
+                    <field name="extractor_voltage" type="NX_FLOAT" units="NX_VOLTAGE"/>
+                    <group name="objective_lens" type="NXlens_em">
+                        <doc>
+                             list of electromagnetic lenses of the microscope column.
```
How can this be a list if there is only one of them?

lukaspie commented on contributed_definitions/NXmpes_peem.nxdl.xml:
```diff
@@ -0,0 +1,202 @@
+                    </group>
+                </group>
+                <!--other EM lenses in column-->
+                <group name="electron_gun" type="NXebeam_column" optional="true">
```
Could be added to `NXmpes` as well

PeterC-DLS replied:
Done

lukaspie commented on contributed_definitions/NXmpes_peem.nxdl.xml:
```diff
@@ -0,0 +1,202 @@
+                <group type="NXenergydispersion" optional="true">
+                    <field name="energy_interval" type="NX_FLOAT" optional="true" units="NX_ENERGY">
+                        <doc>
+                             only in dispersive mode
```
Could benefit from a bit more explanation?

Also, maybe we could add that to the base class?

lukaspie commented on contributed_definitions/NXmpes_peem.nxdl.xml:
```diff
@@ -0,0 +1,202 @@
+                    </group>
+                    <field name="dark_image" type="NX_NUMBER">
+                        <doc>
+                             external link to a dark image.
```
Should this be a link to an external file?

lukaspie commented on contributed_definitions/NXmpes_peem.nxdl.xml:
```diff
@@ -0,0 +1,202 @@
+                        </doc>
+                    </field>
+                </group>
+                <group type="NXdetector">
```
We are using `NXelectron_detector` in `NXmpes` now

lukaspie commented on contributed_definitions/NXmpes_peem.nxdl.xml:
```diff
@@ -0,0 +1,202 @@
+                             external link to a flat field image.
+                        </doc>
+                    </field>
+                    <field name="data" type="NX_NUMBER"/>
```
We have the [`raw_data`](https://fairmat-nfdi.github.io/nexus_definitions/classes/applications/NXmpes.html#nxmpes-entry-instrument-electronanalyzer-electron-detector-raw-data-group) group in `NXmpes` which is an `NXdata`. Would it be okay to use that one here?

PeterC-DLS replied:
Done

lukaspie commented on contributed_definitions/NXmpes_peem.nxdl.xml:
```diff
@@ -0,0 +1,202 @@
+                    </field>
+                    <field name="data" type="NX_NUMBER"/>
+                </group>
+                <group type="NXmanipulator">
+                    <doc>
+                         Manipulator for positioning and rotating of the sample.
+                    </doc>
+                    <group type="NXpositioner">
+                        <doc>
+                             One of a group of sample positioners
+                        </doc>
+                    </group>
```
Not sure if we need these, they are part of `NXmpes` already (`NXpositioner` is part of the `NXmanipulator` base class)

lukaspie commented on contributed_definitions/NXmpes_peem.nxdl.xml:
```diff
@@ -0,0 +1,202 @@
+                        </doc>
+                    </group>
+                    <!--other NXpositioners-->
+                    <field name="sample_bias" type="NX_FLOAT" units="NX_VOLTAGE">
```
There is [`sample_bias_potentiostat`](https://fairmat-nfdi.github.io/nexus_definitions/classes/base_classes/NXmanipulator.html#nxmanipulator-sample-bias-potentiostat-group) in the base class, we should probably use that

lukaspie commented on contributed_definitions/NXmpes_peem.nxdl.xml:
```diff
@@ -0,0 +1,202 @@
+                </group>
+            </group>
+        </group>
+        <group name="geometries" type="NXcoordinate_system_set">
```
This grouping does not exist anymore. 

We have `[/NXmpes/ENTRY/COORDINATE_SYSTEM-group`](https://fairmat-nfdi.github.io/nexus_definitions/classes/applications/NXmpes.html#nxmpes-entry-coordinate-system-group), so we could probably just use that or specialize it here.

lukaspie reviewed (changes requested):
@PeterC-DLS I would like to come back to this proposal at some point. We also have some PEEM users in FAIRmat and it would be helpful to offer such a description to them as well.

I think what you suggested is already very reasonable, we just need to adjust to the latest developments.

My suggestion would be that we adjust the definition a bit to fit to the current NeXus standard and then we add this definition as a contributed one in the NIAC repo. Then we still don't have it completely fixed and if there is more feedback from the community going forward, we can still adjust a bit. What do you think?

PeterC-DLS commented on contributed_definitions/NXmpes_peem.nxdl.xml:
```diff
@@ -0,0 +1,202 @@
+        </field>
+        <group type="NXmonitor" recommended="true">
+            <doc>
+                 contains field named in the **flux** attribute of :ref:`NXbeam`
```
Ok, can do

PeterC-DLS commented:
I will re-examine the recently merged `NXmpes`, rebase this PR and check with our beamline staff and others on updating the commented fields.

lukaspie commented:
> I will re-examine the recently merged `NXmpes`, rebase this PR and check with our beamline staff and others on updating the commented fields.

Perfect thanks! I will also try to gather some feedback from our users.

cc @rettigl

PeterC-DLS committed:
Update PEEM definition with changes in main branch

lukaspie commented on contributed_definitions/NXpeem.nxdl.xml:
```diff
@@ -0,0 +1,203 @@
+                </group>
+                <!--other EM lenses in column-->
+                <group name="electron_gun" type="NXebeam_column" optional="true">
+                    <group name="field_aperture" type="NXaperture_em"/>
```
```suggestion
                    <group name="field_aperture" type="NXaperture"/>
```
Base class does not exist anymore.

PeterC-DLS committed:
Update base class names

PeterC-DLS commented on contributed_definitions/NXmpes_peem.nxdl.xml:
```diff
@@ -0,0 +1,202 @@
+                </field>
+            </group>
+            <group name="beam_probe" type="NXbeam" optional="true">
+                <attribute name="flux"/>
```
I don't think this should be an attribute!

PeterC-DLS replied:
Switched to a field

PeterC-DLS committed:
Update from review

Push an electron gun to NXmpes, clarify that external links are to fields in external files,
fix flux, rename items as raw_data and sample_bias_potentiostat

PeterC-DLS committed:
Document that beam flux should be a link to a monitor's data field

Also, state that an associated_beam can be the name of an NXbeam group
in the enclosing instrument group.

PeterC-DLS commented on contributed_definitions/NXmpes_peem.nxdl.xml:
```diff
@@ -0,0 +1,202 @@
+            <group name="source_probe" type="NXsource" optional="true">
+                <field name="associated_beam" optional="true">
+                    <enumeration>
+                        <item value="beam_probe"/>
```
Done

PeterC-DLS committed:
Update PR with round-trip of yaml/nxdl generation

PeterC-DLS committed:
Fix truncated sentence in application definitions listing

PeterC-DLS committed:
Fix type in beam probe direction

PeterC-DLS committed:
Fix internal referencing
